sample-viewmodel/src/main/java/com/example/molecule/viewmodel/MoleculeViewModel.kt

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2022 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.molecule.viewmodel
+
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.platform.AndroidUiDispatcher
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import app.cash.molecule.RecompositionClock.ContextClock
+import app.cash.molecule.launchMolecule
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.StateFlow
+
+abstract class MoleculeViewModel<Event, Model> : ViewModel() {
```

Comment:
What if instead this:

```kotlin
override suspend fun show(state: MyVmApi.(MyUiState)->Unit) =
  launchMolecule {
    val uiState = remember { mutableStateOf(initialMyUiState) }
    val api = object : MyVmApi {
      // a bunch of methods that update uiState or do other stuff
    }

    // ??? maybe we don't need `key` here???
    key(uiState) {
      api.state(uiState)
    }
  }
```

Reply:
Could also preserve `uiState` by making it a property on `MoleculeViewModel`

Reply:
OK, back from holiday and ready to look into this again. What'cha mean? Could you flesh out this approach? I'm having a hard time extrapolating what this would actually look like.

Reply:
Boy, it's been a while since I wrote this. But IIRC....

The idea is to get rid of `Flow` and have events just be methods. The implementations of those events just modify state. This is the "right" way to use a `ViewModel`, after all.

We can do that by defining the events as an API surface. So `MyVmApi` could also be called `MyEventSink` or something like that.

Reply:
After toying around with it, I think that might be adding too much magic? The advantage of
```kotlin
@Composable
protected abstract fun models(events: Flow<Event>): Model
```
is that it's easy to reason about, and provides a simple API surface for producing an initial model and subsequent updates.

Happy to pair on this though if you have a clear vision in mind that you think would be better.